### PR TITLE
#377 - have the ability to know and capture all fields in a data fetcher

### PR DESCRIPTION
**Pull request by bbakerman**
#377  - have the ability to know and capture all fields in a data fetcher

This tweaks DataFetcher so it has the ExecutionContext available and FieldCollector to have a better public API

### CODE REVIEW

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 1]
```diff
@@ -140,12 +140,7 @@ protected ExecutionResult completeValue(ExecutionContext executionContext, Execu
             resolvedType = (GraphQLObjectType) fieldType;
         }
 
-        Map<String, List<Field>> subFields = new LinkedHashMap<>();
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java]: Notice how visitedFragments is NEVER used.  Thats why I took it out of the public method signature

File: src/main/java/graphql/execution/FieldCollector.java [hunk 2]
```diff
@@ -21,7 +35,45 @@ public FieldCollector() {
...
+     */
+    public Map<String, List<Field>> collectFields(ExecutionContext executionContext, GraphQLObjectType objectType, SelectionSet selectionSet) {
+        Map<String, List<Field>> subFields = new LinkedHashMap<>();
+        List<String> visitedFragments = new ArrayList<>();
+        this.collectFields(executionContext, objectType, selectionSet, visitedFragments, subFields);
+        return subFields;
+    }
+
+
+    private void collectFields(ExecutionContext executionContext, GraphQLObjectType type, SelectionSet selectionSet, List<String> visitedFragments, Map<String, List<Field>> fields) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/FieldCollector.java]: I can make this public again but I figured 3.x and breaking changes and its unlikely to break ANYONE really

File: src/main/java/graphql/schema/DataFetchingEnvironment.java [hunk 3]
```diff
@@ -81,4 +82,9 @@
      * @return the {@link ExecutionId} for the current operation
      */
     ExecutionId getExecutionId();
+
+    /**
+     * @return the {@link ExecutionContext} for the current operation
+     */
+    ExecutionContext getExecutionContext();
```
Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: To use FieldCollector in a DataFetcher, I need the ExecutionContext.

**Comment by bbakerman** [comment]
@kaqqao - this helps solve your selection problem.  I think all the parts are now in place

File: src/main/java/graphql/schema/DataFetchingEnvironment.java (hunk 3 above)
Comment (andimarek) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: @bbakerman Does this mean ExecutionContext is now part of the public API?

**Comment by kaqqao** [comment]
Commented on [#377](https://github.com/graphql-java/graphql-java/issues/377#issuecomment-299162821) about the idea I had for `FieldCollector`. Basically, I was thinking it could be created once per execution (with all variables and fragment set) and passed into every `DatafetchingEnvironment`. Maybe even completely hidden from the user and only exposed via new methods on `DatafetchingEnvironment`.

**Comment by bbakerman** [comment]
>  Basically, I was thinking it could be created once per execution (with all variables and fragment set) and passed into every DatafetchingEnvironment. 

Thats brilliant - yes that what we should do.  Have an API wrapper for field collector and leave all the plumbing to the internals.

**Comment by andimarek** [comment]
I agree with @bbakerman sound good. 👍

**Comment by bbakerman** [comment]
OK so I have created a DataFetcherSelectionSet interface so we can hide the details of how to get the selection set of fields from consumers.

I used a supplier interface for DataFetchingFieldSelectionSet so that it can be lazy.  If you dont want the "selection set" of fields then it wont evaluate them.

**Comment by andimarek** [comment]
@bbakerman why is this a breaking change?

**Comment by bbakerman** [comment]
Its a minor breaking change in that I refactored FieldCollector.  Fiven that we dont expect consumers to consume FieldCollector its debtateable.

**Pull request merged**